File: 2020/day07/puzzle7.py
import re

file = open('puzzle7.in')

rules = {}
my_color = 'shiny gold'

for line in file:
  rule = line.strip()
  [container, content_list] = rule.split(' bags contain ')
  if content_list == 'no other bags.':
    rules[container] = []
  else:
    contents = content_list.split(', ')
    parsed_contents = []

    for content in contents:
      [amount, color] = content.split(' ', 1)
      parsed_contents.append({ 'amount': int(amount), 'color': re.sub(r' bags?\.?$', '', color) })
    rules[container] = parsed_contents

def search_containers(search_color):
  container_colors = set()

  for color in rules:
    for content in rules[color]:
      if content['color'] == search_color:
        container_colors.add(color);
        container_colors.update(search_containers(color))

  return container_colors

def part1():
  return len(search_containers(my_color))

def count_contents(search_color):
  content_count = 0

  for content in rules[search_color]:
    sub_content_count = count_contents(content['color'])
    content_count += content['amount'] * (sub_content_count + 1)

  return content_count

def part2():
  return count_contents(my_color)

print(part1())
print(part2())

File: 2020/day07/puzzle7.ts
const input = await Deno.readTextFile("puzzle7.in");

type Content = {
  amount: number;
  color: string;
};

type Rule = {
  container: string;
  contents: Content[];
};

const myColor = "shiny gold";

function parseContent(content: string) {
  const [amount, ...colorParts] = content.replace(/ bags?\.?/, "").split(" ");
  return { amount: parseInt(amount), color: colorParts.join(" ") };
}

function parseRule(rule: string) {
  const [container, contentList] = rule.split(" bags contain ");
  const contents =
    contentList === "no other bags." ? [] : contentList.split(", ").map(parseContent);
  return { container, contents };
}

const rules = input.split("\n").map(parseRule);

function searchContainers(searchColor: string) {
  const containers: string[] = rules
    .filter((rule: Rule) => rule.contents.some((content: Content) => content.color === searchColor))
    .map((rule: Rule) => rule.container)
    .map((color: string) => [color, ...searchContainers(color)])
    .flat();

  return [...new Set(containers)];
}

function part1() {
  return searchContainers(myColor).length;
}

function countContents(searchColor: string): number {
  const colorRule = rules.find((rule: Rule) => rule.container === searchColor);

  if (!colorRule) {
    return 0;
  }

  return colorRule.contents
    .map((content: Content) => content.amount * (countContents(content.color) + 1))
    .reduce((sum: number, contentCount: number) => sum + contentCount, 0);
}

function part2() {
  return countContents(myColor);
}

console.log(part1());
console.log(part2());
